Add Fastly Object Storage guide to documentation. (#1024)
fastly/terraform-provider-fastly: templates/guides/fastly_object_storage.md

---
page_title: Managing content in Fastly Object Storage
subcategory: "Guides"
---
# Managing content in Fastly Object Storage

Content in Fastly Object Storage (buckets and objects) can be managed
using a combination of this provider (Fastly) and the Amazon Web
Services (AWS) provider, since Fastly Object Storage provides an AWS
S3-compatible API.

## Example Configuration

Two HCL files are required. The first, `main.tf`:

```terraform
terraform {
  required_providers {
    fastly = {
      source  = "fastly/fastly"
      version = "~> 7.1.0"
    }
  }
}

resource "fastly_object_storage_access_keys" "main" {
  description = "FOS Key"
  permission  = "read-write-admin"
}

module "fos" {
  source = "./fos"
  access_key_id = fastly_object_storage_access_keys.main.id
  secret_key    = fastly_object_storage_access_keys.main.secret_key
}

> ```

Reviewer: It feels like there's something missing here, either a missing word (which requires correcting punctuation), or some sort of followup statement to complete the second sentence (which requires text after the code block). The `fos/main.tf` code block below this states that it need to be placed and named specifically. Perhaps that was the plan for `main.tf` as well?

Author: Yes, that is the case, this file must be named `main.tf` and that's standard practice for creating a Terraform configuration (it's the only required file in any configuration).

> The second will need to be placed in a directory named `fos`, and
> named `fos/main.tf`:

Reviewer: The comma here is unnecessary:

```suggestion
The second will need to be placed in a directory named `fos` and
named `fos/main.tf`:
```

```terraform
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "= 6.0"
    }
  }
}

provider "aws" {
  access_key = var.access_key_id
  secret_key = var.secret_key
  region     = "us-east"

  s3_use_path_style                  = true
  skip_credentials_validation        = true
  skip_metadata_api_check            = true
  skip_region_validation             = true
  skip_requesting_account_id         = true

  endpoints {
    s3 = "https://us-east.object.fastlystorage.app"
  }
}

resource "aws_s3_bucket" "main" {
  bucket = "my-test-bucket-123"

  # Fastly object storage uses different region names
  lifecycle {
    ignore_changes = [region]
  }
}

```

> Note: This example uses the `us-east` region of Fastly Object Storage;
> if you wish to use a different region, ensure that the proper region
> code is included in the `endpoints` block above.

Reviewer: From a preferred words and punctuation standpoint, we suggest updating this as follows:

```suggestion
Note: This example uses the `us-east` region of Fastly Object Storage.
To use a different region, ensure that the proper region
code is included in the `endpoints` block above.
```

## Getting Started

> With the example files in place, you'll need to initialize Terraform
> and obtain Fastly Object Storage credentials.

Reviewer: This is correct but could benefit from a direct call to action for the reader to do something. Consider this as an option:

```suggestion
With the example files in place, you'll need to initialize Terraform
and obtain Fastly Object Storage credentials. To do this, run the following
commands:
```

```bash
export FASTLY_API_KEY=<your Fastly API key here>
terraform init
terraform apply -target=fastly_object_storage_access_keys.main
```

> Note: Terraform will issue a warning because the `-target` option is
> used. This usage of that option is safe.

Reviewer: Not a blocker. We tend to use "use" instead of "usage" to sound more friendly.

```suggestion
Note: Terraform will issue a warning because the `-target` option is
used. This use of that option is safe.
```

This step will connect to the Fastly API using the Fastly Terraform
provider and obtain a set of Fastly Object Storage credentials. Those
credentials will be stored in the Terraform state files (or other

> state storage), they will not be displayed.

Reviewer: There's a small punctuation or rewrite fix that needs to be made here to correct the run-on sentence. Here's one option:

```suggestion
This step will connect to the Fastly API using the Fastly Terraform
provider and obtain a set of Fastly Object Storage credentials. Those
credentials will be stored in the Terraform state files (or other
state storage) and will not be displayed.
```

> This initial step is necessary because the credentials are required by
> the AWS Terraform provider, and if Terraform attempts to apply the
> entire configuration the AWS provider will report an error because the
> credentials are missing.

Reviewer: Not a blocker, but to shorten this for clarity, you could split this into two separate sentences to make it easier to read (especially for non-English speakers).

```suggestion
This initial step is necessary because the credentials are required by
the AWS Terraform provider. If Terraform attempts to apply the
entire configuration without them, the AWS provider will report an error because the
credentials are missing.
```

## Completing the Process

> With the credentials obtained, a normal Terraform `apply` step can be
> used to create the remaining infrastructure; in this case a bucket in
> Fastly Object Storage named `my-test-bucket-123`. Terraform will pass
> the Fastly Object Storage credentials to the AWS Terraform provider so
> that it can use them to authenticate its API interactions with the
> Fastly Object Storage system.

Reviewer: Is "normal" the correct word to consider here or would "standard" work better? If the intent is to highlight that there can be abnormal things, then it's fine. I suspect, however, we're aiming for the sense of "Hey, nothing complicated. Just do this like normal.", in which case the following would be more correct:

```suggestion
With the credentials obtained, a standard Terraform `apply` step can be
used to create the remaining infrastructure; in this case, a bucket in
Fastly Object Storage named `my-test-bucket-123`. Terraform will pass
the Fastly Object Storage credentials to the AWS Terraform provider so
that it can use them to authenticate its API interactions with the
Fastly Object Storage system.
```
